api/src/wfl/module/covid.clj

```diff
@@ -93,6 +93,98 @@
   (fn [tx workload] (:sink_type workload)))
 
+;; validations
+(defmulti throw-when-malformed-source-request!
+  "Validate the source of a `request`"
+  (fn [source] (:name source)))
+
+(defmulti throw-when-malformed-executor-request!
+  "Validate the executor of a `request`"
+  (fn [executor] (:name executor)))
+
+(defmulti throw-when-malformed-sink-request!
+  "Validate the sink of a `request`"
+  (fn [sink] (:name sink)))
+
+;; Workload Functions
+(def ^:private dataset-table-name "flowcells")
+(def ^:private dataset-column-name "run_date")
+(def ^:private source-name "Terra DataRepo")
+(def ^:private executor-name "Terra")
+(def ^:private sink-name "Terra Workspace")
+
+(defn ^:private add-workload-record
+  "Use `tx` to create a workload `record` for `request` and return the id of the
+   new workload."
+  [tx request]
+  (letfn [(combine-labels [labels]
+            (->> (mapv request [:pipeline :project])
+                 (map str ["pipeline:" "project:"])
+                 (concat labels)
+                 set
+                 sort
+                 vec))]
+    (-> (update request :labels combine-labels)
+        (select-keys [:creator :watchers :labels :project])
+        (merge (select-keys (wfl/get-the-version) [:commit :version]))
+        (assoc :executor "" :output "" :release "" :wdl "" :uuid (UUID/randomUUID))
+        (->> (jdbc/insert! tx :workload) first :id))))
+
+(defn throw-unless-column-exists
+  "Throw or return the column from `table`"
+  [dataset table]
+  (let [[result & more :as all] (-> table
+                                    (get-in [:columns])
+                                    (->> (filter (comp #{dataset-column-name} :name))))]
+    (when-not result
+      (throw (ex-info "No column with name" {:name dataset-column-name :dataset dataset})))
+    result))
+
+(defn throw-unless-table-exists
+  "Throw or return the table from `dataset`"
+  [dataset]
+  (let [[result & more :as all] (-> dataset
+                                    (get-in [:schema :tables])
+                                    (->> (filter (comp #{dataset-table-name} :name))))]
+    (when-not result
+      (throw (ex-info "No table with name" {:name dataset-table-name :dataset dataset})))
+    (when result
+      (throw-unless-column-exists dataset result))
+    result))
+
+(defn verify-data-repo-source!
+  "Verify that the `dataset` exists and that the WFL has the necessary permissions to read it"
+  [{:keys [name dataset] :as source}]
+  (when-not (some? dataset)
+    (throw (ex-info "Source is nil" {:source source})))
+  (when-not (= (:name source) source-name)
+    (throw (ex-info "Unknown Source" {:source source})))
+  (-> (datarepo/dataset (:dataset source))
+      (throw-unless-table-exists)))
+
+(defn verify-terra-executor!
+  "Verify the method-configuration exists."
+  [{:keys [name method_configuration] :as executor}]
+  (when-not (= (:name executor) executor-name)
+    (throw (ex-info "Unknown Executor" {:executor executor})))
```

Comment:
That should be thrown by the multimethod's `:default` implementation.

Reply:
Updated